metric_learn/base_metric.py

```diff
@@ -161,6 +161,68 @@ def transform(self, X):
     """
 
 
+class BilinearMixin(BaseMetricLearner, metaclass=ABCMeta):
+
+  def score_pairs(self, pairs):
+      r"""
+      Parameters
+      ----------
+      pairs : array-like, shape=(n_pairs, 2, n_features) or (n_pairs, 2)
+        3D Array of pairs to score, with each row corresponding to two points,
+        for 2D array of indices of pairs if the metric learner uses a
+        preprocessor.
+
+      Returns
+      -------
+      scores : `numpy.ndarray` of shape=(n_pairs,)
+        The learned Mahalanobis distance for every pair.
+      """
+      check_is_fitted(self, ['preprocessor_', 'components_'])
+      pairs = check_input(pairs, type_of_inputs='tuples',
+                          preprocessor=self.preprocessor_,
+                          estimator=self, tuple_size=2)
+      # Note: For bilinear order matters, dist(a,b) != dist(b,a)
+      # We always choose first pair first, then second pair
+      # (In contrast with Mahalanobis implementation)
+      # I dont know wich implementation performs better
+      return np.diagonal(np.dot(
+                      np.dot(pairs[:, 0, :], self.components_),
+                      pairs[:, 1, :].T))
+      return np.array([np.dot(np.dot(u.T, self.components_), v)
+                      for u, v in zip(pairs[:, 0, :], pairs[:, 1, :])])
+
+  def get_metric(self):
+      check_is_fitted(self, 'components_')
+      components = self.components_.copy()
+
+      def metric_fun(u, v):
+          """This function computes the metric between u and v, according to the
+          previously learned metric.
+
+          Parameters
+          ----------
+          u : array-like, shape=(n_features,)
+            The first point involved in the distance computation.
+
+          v : array-like, shape=(n_features,)
+            The second point involved in the distance computation.
+
+          Returns
+          -------
+          distance : float
+            The distance between u and v according to the new metric.
+          """
```

Comment:
comments should be updated to reflect that we do not have a distance anymore but a similarity

Reply:
All comments have been updated stating this is in fact a similarity.

The only thing that makes me noise is the `get_metric` method. It might be `get_similarity`, but we can't do that as `get_metric` is inherited from `BaseMetricLearner`. But documentation should be enough.

Reply:
Yes, this is fine I think